File: src/main/java/io/github/archipelagominecraft/core/loaders/fabric/AdvancementEarnEvent.java
//? if fabric {
/*package io.github.archipelagominecraft.core.loaders.fabric;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.advancements.AdvancementHolder;
import net.minecraft.server.level.ServerPlayer;

public interface AdvancementEarnEvent {
    Event<AdvancementEarnEvent> EVENT = EventFactory.createArrayBacked(AdvancementEarnEvent.class,
        (listeners) -> (advancement, player) -> {
            for (AdvancementEarnEvent listener : listeners) {
                listener.onAdvancementEarned(advancement, player);
            }
        });

    void onAdvancementEarned(AdvancementHolder advancementHolder, ServerPlayer player);
}
*///?}

File: src/main/kotlin/io/github/archipelagominecraft/core/vanilla/events/AdvancementEarnEvent.kt
package io.github.archipelagominecraft.core.vanilla.events

import io.github.archipelagominecraft.core.vanilla.compat.ResourceLocation
import io.github.archipelagominecraft.core.vanilla.compat.ServerPlayer
import net.minecraft.advancements.Advancement

data class AdvancementEarnEvent(
    val advancementHolder: Advancement,
    val id: ResourceLocation,
    val player: ServerPlayer
) {
    companion object {
        @JvmField
        val EVENT: ArchipelagoEvent<AdvancementEarnEvent> = ArchipelagoEvent()
    }
}
